File: source/105.从前序与中序遍历序列构造二叉树.cpp
/*
 * @lc app=leetcode.cn id=105 lang=cpp
 *
 * [105] 从前序与中序遍历序列构造二叉树
 */
#include "cpp_includes.h"
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int, int> inHashmap;
        for (int i = 0; i < inorder.size(); i++) {
            inHashmap[inorder[i]] = i;
        }

        int preorderIndex = 0;
        return buildTree(preorder, inHashmap, 0, inorder.size()-1, preorderIndex);
    }

    TreeNode* buildTree(vector<int>& preorder, unordered_map<int,int>& inhashMap, int is, int ie, int& ps) {
        if (ps == preorder.size() || is > ie) return nullptr;

        auto node = new TreeNode(preorder[ps]);
        int inorderIndex = inhashMap[preorder[ps]];
        ps++;

        node->left = buildTree(preorder, inhashMap, is, inorderIndex-1, ps);
        node->right = buildTree(preorder, inhashMap, inorderIndex+1, ie, ps);

        return node;
    }
};
// @lc code=end

File: source/105.从前序与中序遍历序列构造二叉树.cs
/*
 * @lc app=leetcode.cn id=105 lang=csharp
 *
 * [105] 从前序与中序遍历序列构造二叉树
 */
using csharpnaming;
// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution
{
  public TreeNode BuildTree(int[] preorder, int[] inorder)
  {
    var inDict = new Dictionary<int, int>();
    for (int i = 0; i < inorder.Length; i++)
    {
      inDict[inorder[i]] = i;
    }

    int pos = 0;
    return BuildTree(preorder, inDict, 0, inorder.Length - 1, ref pos);
  }

  private TreeNode BuildTree(int[] preorder, Dictionary<int, int> inDict, int ibegin, int iend, ref int pos)
  {
    if (pos == preorder.Length || ibegin > iend) return null;

    var node = new TreeNode(preorder[pos]);
    var i = inDict[preorder[pos]];
    pos++;

    node.left = BuildTree(preorder, inDict, ibegin, i - 1, ref pos);
    node.right = BuildTree(preorder, inDict, i + 1, iend, ref pos);

    return node;
  }
}
// @lc code=end

